storage/src/main/scala/coop/rchain/storage/StorageActions.scala

```diff
@@ -0,0 +1,118 @@
+package coop.rchain.storage
+
+import cats.implicits._
+import coop.rchain.storage.util.ignore
+
+import scala.annotation.tailrec
+
+// TODO(ht): improve docstrings
+trait StorageActions {
+
+  /* Consume */
+
+  @tailrec
+  private[storage] final def findMatchingDataCandidates[C, P, A](
+      data: List[(A, Int)],
+      pattern: P,
+      channel: C)(implicit m: Match[P, A]): Option[(A, C, Int)] =
+    data match {
+      case Nil => None
+      case (a, i) :: as =>
+        m.get(pattern, a) match {
+          case None      => findMatchingDataCandidates(as, pattern, channel)
+          case Some(mat) => Some((mat, channel, i))
+        }
+    }
+
+  private[storage] def extractDataCandidates[C, P, A, K](
+      store: IStore[C, P, A, K],
+      channels: List[C],
+      patterns: List[P])(txn: store.T)(implicit m: Match[P, A]): Option[List[(A, C, Int)]] = {
+    val options = channels.zip(patterns).map {
+      case (c, p) =>
+        val as: List[(A, Int)] = store.getAs(txn, c.pure[List]).zipWithIndex
+        findMatchingDataCandidates(as, p, c)
+    }
+    options.sequence[Option, (A, C, Int)]
+  }
+
+  /** `consume` does the "consume" thing
+    *
+    * @param store
+    * @param channels
+    * @param patterns
+    * @param continuation
+    * @tparam C a type representing a channel
+    * @tparam P a type representing a pattern
+    * @tparam A a type representing a piece of data
+    * @tparam K a type representing a continuation
+    */
+  def consume[C, P, A, K](store: IStore[C, P, A, K],
+                          channels: List[C],
+                          patterns: List[P],
+                          continuation: K)(implicit m: Match[P, A]): Option[(K, List[A])] = {
+    if (channels.length =!= patterns.length)
+      throw new IllegalArgumentException("channels.length must equal patterns.length")
+    store.withTxn(store.createTxnWrite()) { txn =>
+      extractDataCandidates(store, channels, patterns)(txn) match {
+        case None =>
+          store.putK(txn, channels, patterns, continuation)
+          for (c <- channels) store.addJoin(txn, c, channels)
+          None
+        case Some(acis) =>
+          acis.foreach {
+            case (_, c, i) =>
+              store.removeA(txn, c.pure[List], i)
+              store.removeK(txn, channels, i)
```

Comment:
If you get an immediate match, which continuation are you removing from the store?

Reply:
Yep, good point.  Removed.